pip/src/interop.rs

```diff
@@ -0,0 +1,657 @@
+    let kwargs = kwargs.unwrap_or_else(|| PyDict::new_bound(py));
+
+    let name = get_name(&kwargs)?;
+    let search_path = get_search_path(&kwargs)?;
+
+    let fs = create_filesystem_from_py(py, read_file, list_directory, resolve_path, fetch_github);
+    let resolver = ImportResolver::new(fs, PathBuf::from(search_path));
+
+    let program_type = ProgramType::File(name.to_string());
+    let output_semantics = OutputSemantics::ResourceEstimation;
+    let unit = compile_qasm(source, &name, &resolver, program_type, output_semantics)?;
+    let (source_map, _, package, _) = unit.into_tuple();
+    match crate::interop::estimate_qasm3(
+        package.expect("Package must exist when there are no errors"),
+        source_map,
+        job_params,
+    ) {
+        Ok(estimate) => Ok(estimate),
+        Err(errors) if matches!(errors[0], re::Error::Interpreter(_)) => {
+            Err(QSharpError::new_err(format_errors(
+                errors
+                    .into_iter()
+                    .map(|e| match e {
+                        re::Error::Interpreter(e) => e,
+                        re::Error::Estimation(_) => unreachable!(),
+                    })
+                    .collect::<Vec<_>>(),
+            )))
+        }
+        Err(errors) => Err(QSharpError::new_err(
+            errors
+                .into_iter()
+                .map(|e| match e {
+                    re::Error::Estimation(e) => e.to_string(),
+                    re::Error::Interpreter(_) => unreachable!(),
+                })
+                .collect::<Vec<_>>()
+                .join("\n"),
+        )),
+    }
+}
+
+/// This call while exported is not intended to be used directly by the user.
+/// It is intended to be used by the Python wrapper which will handle the
+/// callbacks and other Python specific details.
+#[pyfunction]
+#[allow(clippy::too_many_arguments)]
+#[pyo3(
+    signature = (source, read_file, list_directory, resolve_path, fetch_github, **kwargs)
+)]
+pub(crate) fn compile_qasm3_to_qir(
+    py: Python,
+    source: &str,
+    read_file: Option<PyObject>,
+    list_directory: Option<PyObject>,
+    resolve_path: Option<PyObject>,
+    fetch_github: Option<PyObject>,
+    kwargs: Option<Bound<'_, PyDict>>,
+) -> PyResult<String> {
+    let kwargs = kwargs.unwrap_or_else(|| PyDict::new_bound(py));
+
+    let target = get_target_profile(&kwargs)?;
+    let name = get_name(&kwargs)?;
+    let search_path = get_search_path(&kwargs)?;
+
+    let fs = create_filesystem_from_py(py, read_file, list_directory, resolve_path, fetch_github);
+    let resolver = ImportResolver::new(fs, PathBuf::from(search_path));
+
+    let program_type = ProgramType::File(name.to_string());
+    let (package, source_map, signature) = compile_qasm_enriching_errors(
+        source,
+        &name,
+        &resolver,
+        program_type,
+        OutputSemantics::Qiskit,
+        false,
+    )?;
+
+    let package_type = PackageType::Lib;
```

Comment:
am I correct in understanding that all `qasm3` to `qir` compilations will be `PackageType::Lib`, and `LanguageFeatures::default()`? Do we want additional configurability here?

Reply:
For this code path we don't know the entry expr until after compilation so we can't be exe. When running in interactive, we'll just be adding to the run state, so we will use `interperter.qir(...)` path.